codegen-winter/src/air/boundary_constraints.rs

```diff
@@ -109,28 +108,48 @@ impl Codegen for BoundaryExpr {
             }
             Self::Add(lhs, rhs) => {
                 format!(
-                    "({}) + ({})",
+                    "{} + {}",
                     lhs.to_string(is_aux_constraint),
                     rhs.to_string(is_aux_constraint)
                 )
             }
             Self::Sub(lhs, rhs) => {
-                format!(
-                    "({}) - ({})",
-                    lhs.to_string(is_aux_constraint),
+                let rhs = if is_arithmetic_expr(rhs) {
+                    format!("({})", rhs.to_string(is_aux_constraint))
+                } else {
                     rhs.to_string(is_aux_constraint)
-                )
+                };
+                format!("{} - {}", lhs.to_string(is_aux_constraint), rhs)
             }
             Self::Mul(lhs, rhs) => {
-                format!(
-                    "({}) * ({})",
-                    lhs.to_string(is_aux_constraint),
+                let lhs = if is_arithmetic_expr(lhs) {
+                    format!("({})", lhs.to_string(is_aux_constraint))
+                } else {
+                    lhs.to_string(is_aux_constraint)
+                };
+                let rhs = if is_arithmetic_expr(rhs) {
+                    format!("({})", rhs.to_string(is_aux_constraint))
+                } else {
                     rhs.to_string(is_aux_constraint)
-                )
+                };
+                format!("{} * {}", lhs, rhs)
             }
             Self::Exp(lhs, rhs) => {
-                format!("({}).exp({})", lhs.to_string(is_aux_constraint), rhs)
+                let lhs = if is_arithmetic_expr(lhs) {
+                    format!("({})", lhs.to_string(is_aux_constraint))
+                } else {
+                    lhs.to_string(is_aux_constraint)
+                };
+                format!("{}.exp({})", lhs, rhs)
             }
         }
     }
 }
+
+/// Checks whether the boundary expression is an arithmetic operation.
+fn is_arithmetic_expr(boundary_expr: &BoundaryExpr) -> bool {
+    matches!(
+        boundary_expr,
+        BoundaryExpr::Add(_, _) | BoundaryExpr::Sub(_, _) | BoundaryExpr::Mul(_, _)
+    )
+}
```

Comment:
This method is going to produce extra parentheses. For example, `a * b * c` will result in `a * (b * c)` unnecessarily. I think a better way to do this would be to think about operator precedence (or binding power), and use parens when the sub expression has a lower precedence. We could probably do something elegant like reverse Pratt parsing here.

Reply:
Got it. Thank you. Will look into this later this week.